README.md
# spur.py: Run commands and manipulate files locally or over SSH using the same interface

To run echo locally:

```python
import spur

shell = spur.LocalShell()
result = shell.run(["echo", "-n", "hello"])
print result.output # prints hello
```

Executing the same command over SSH uses the same interface -- the only difference
is how the shell is created:

```python
import spur

shell = spur.SshShell(hostname="localhost", username="bob", password="password1")
with shell:
    result = shell.run(["echo", "-n", "hello"])
print result.output # prints hello
```

## Installation

```$ pip install spur```

## Shell constructors

### LocalShell

Takes no arguments:

```python
spur.LocalShell()
```

### SshShell

Requires a hostname. Also requires some combination of a username, password
and private key, as necessary to authenticate:

```python
# Use a password
spur.SshShell(
    hostname="localhost",
    username="bob",
    password="password1"
)
# Use a private key
spur.SshShell(
    hostname="localhost",
    username="bob",
    private_key_file="path/to/private.key"
)
# Use a port other than 22
spur.SshShell(
    hostname="localhost",
    port=50022,
    username="bob",
    password="password1"
)
```

Optional arguments:

* `connect_timeout` -- a timeout in seconds for establishing an SSH connection.
  Defaults to 60 (one minute).
  
* `missing_host_key` -- by default, an error is raised when a host key is
  missing. One of the following values can be used to change the behaviour
  when a host key is missing:
  - `spur.ssh.MissingHostKey.raise_error` -- raise an error
  - `spur.ssh.MissingHostKey.warn` -- accept the host key and log a warning
  - `spur.ssh.MissingHostKey.accept` -- accept the host key

## Shell interface

### run(command, cwd, update_env, store_pid, allow_error, stdout, stderr)

Run a command and wait for it to complete. The command is expected to be a list
of strings. Returns an instance of `ExecutionResult`.

```python
result = shell.run(["echo", "-n", "hello"])
print result.output # prints hello
```

Note that arguments are passed without any shell expansion. For instance,
`shell.run(["echo", "$PATH"])` will print the literal string `$PATH` rather
than the value of the environment variable `$PATH`.

Raises `spur.NoSuchCommandError` if trying to execute a non-existent command.

Optional arguments:

* `cwd` -- change the current directory to this value before executing the
  command.
* `update_env` -- a `dict` containing environment variables to be set before
  running the command. If there's an existing environment variable with the same
  name, it will be overwritten. Otherwise, it is unchanged.
* `store_pid` -- if set to `True` when calling `spawn`, store the process id of
  the spawned process as the attribute `pid` on the returned process object. Has
  no effect when calling `run`.
* `allow_error` -- `False` by default. If `False`, an exception is raised if
  the return code of the command is anything but 0. If `True`, a result is
  returned irrespective of return code.
* `stdout` -- if not `None`, anything the command prints to standard output
  during its execution will also be written to `stdout` using `stdout.write`.
* `stderr` -- if not `None`, anything the command prints to standard error
  during its execution will also be written to `stderr` using `stderr.write`.

`shell.run(*args, **kwargs)` should behave similarly to
`shell.spawn(*args, **kwargs).wait_for_result()`

### spawn(command, cwd, update_env, store_pid, allow_error, stdout, stderr)

Behaves the same as `run` except that `spawn` immediately returns an object
representing the running process.

Raises `spur.NoSuchCommandError` if trying to execute a non-existent command.

### open(path, mode="r")

Open the file at `path`. Returns a file-like object.

## Environmental Variables on Remote Machine

To get environmental variables already set on the remote machine, use:

env_variable = shell.env[{ENVIRONMENTAL_VARIABLE}]


```python
env_variable = shell.env['HOME']
```

To set environmental variables, use:

```python
result = shell.env.setenv(name, value)
```

## Process interface

Returned by calls to `shell.spawn`. Has the following attributes:

* `pid` -- the process ID of the process. Only available if `store_pid` was
  set to `True` when calling `spawn`.

Has the following methods:

* `is_running()` -- return `True` if the process is still running, `False`
  otherwise.
* `stdin_write(value)` -- write `value` to the standard input of the process.
* `wait_for_result()` -- wait for the process to exit, and then return an
  instance of `ExecutionResult`. Will raise `RunProcessError` if the return
  code is not zero and `shell.spawn` was not called with `allow_error=True`.
* `send_signal(signal)` -- sends the process the signal `signal`. Only available
  if `store_pid` was set to `True` when calling `spawn`.

## Classes

### ExecutionResult

`ExecutionResult` has the following properties:

* `return_code` -- the return code of the command
* `output` -- a string containing the result of capturing stdout
* `stderr_output` -- a string containing the result of capturing stdout

It also has the following methods:

* `to_error()` -- return the corresponding RunProcessError. This is useful if
  you want to conditionally raise RunProcessError, for instance:
  
```python
result = shell.run(["some-command"], allow_error=True)
if result.return_code > 4:
    raise result.to_error()
```

### RunProcessError

A subclass of `RuntimeError` with the same properties as `ExecutionResult`:

* `return_code` -- the return code of the command
* `output` -- a string containing the result of capturing stdout
* `stderr_output` -- a string containing the result of capturing stdout

### NoSuchCommandError

`NoSuchCommandError` has the following properties:

* `command` -- the command that could not be found

## API stability

Using the the terminology from [Semantic Versioning][semver], if the version
of spur is X.Y.Z, then X is the major version, Y is the minor version, and Z is
the patch version.

[semver]: http://semver.org/spec/v1.0.0.html

While the major version is 0, incrementing the patch version indicates a
backwards compatible change. For instance, if you're using 0.3.1, then it should
be safe to upgrade to 0.3.2.

Incrementing the minor version indicates a change in the API. This means that
any code using previous minor versions of spur may need updating before it can
use the current minor version.

### Undocumented features

Some features are undocumented, and should be considered experimental. Use them
at your own risk. They may not behave correctly, and their behaviour and
interface may change at any time.

## Troubleshooting

### I get the error "Connection refused" when trying to connect to a virtual machine using a forwarded port on `localhost`

Try using `"127.0.0.1"` instead of `"localhost"` as the hostname.

### I get the error "Connection refused" when trying to execute commands over SSH

Try connecting to the machine using SSH on the command line with the same
settings. For instance, if you're using the code:

```python
shell = spur.SshShell(
        hostname="remote",
        port=2222,
        username="bob",
        private_key_file="/home/bob/.ssh/id_rsa"
    )
with shell:
    result = shell.run(["echo", "hello"])
```

Try running:

```
ssh bob@remote -p 2222 -i /home/bob/.ssh/id_rsa
```

If the `ssh` command succeeds, make sure that the arguments to `ssh.SshShell`
and the `ssh` command are the same. If any of the arguments to `ssh.SshShell`
are dynamically generated, try hard-coding them to make sure they're set to the
values you expect.


CHANGES
# CHANGES

## 0.3.23

* Raise minimum Python version to 3.6.

* Handle FileNotFoundError when using subprocess on Python >= 3.8.

## 0.3.22

* Fix: default connect timeout was not being set correctly.

## 0.3.21

* Add close() method to shells, behaving the same as __exit__().

## 0.3.20

* SshShell: passing None as the port argument will cause port 22 to be used.

* Raise CouldNotChangeDirectoryError when failing to change directory to the cwd
  argument.

## 0.3.19

* Allow spur to be used with Paramiko 2. To quote the Paramiko changelog [1]:

    Replace PyCrypto with the Python Cryptographic Authority (PyCA)
    ‘Cryptography’ library suite. This improves security, installability, and
    performance; adds PyPy support; and much more.

  This should only affect installation requirements for Paramiko: see the
  changelog for more information. Since the API of Paramiko remains unchanged,
  spur should remain compatible with both Paramiko 1 and Paramiko 2.

  Note that Python 3.2 is not compatible with Paramiko 2: you can continue to
  use spur with Python 3.2 and Paramiko 1, although upgrading is recommended.

[1]: http://www.paramiko.org/changelog.html#2.0.0

## 0.3.18

* Add sock argument to SshShell.

## 0.3.17

* When encoding argument is set, decode output before writing to stdout and
  stderr arguments. This changes the behaviour to match the documented
  behaviour: previously, only the output on the result object was decoded.

## 0.3.16

* Remove non-ASCII character from README.rst.

## 0.3.15

* Add encoding argument to spawn and run.

* SshShell: add load_system_host_keys argument.

* LocalShell: don't raise NoSuchCommandError when cwd does not exist.

## 0.3.14

* Raise spur.CommandInitializationError when SshShell fails to read integers
  from stdout before the command is actually run.

## 0.3.13

* Add look_for_private_keys argument to SshShell to allow searching for private
  keys to be disabled.

## 0.3.12

* Add shell_type argument to SshShell to allow better support for minimal
  shells, such as those found on embedded systems.

* Open files in text mode by default. When opening files over SSH, this means
  that files are decoded using the encoding returned by
  locale.getpreferredencoding().

## 0.3.11

* Add support for platforms that don't support the pty module, such as Windows.

## 0.3.10

* SshShell: Use "which" if "command -v" is not available. Fixes GitHub issue #15:

    https://github.com/mwilliamson/spur.py/issues/15

## 0.3.9

* Treat output as bytes rather than attempting to decode to string when
  generating RunProcessError.message. Fixes GitHub issue #13:

    https://github.com/mwilliamson/spur.py/pull/13

* Support unicode commands over SSH.

## 0.3.8

* Add full support for Python 3.

## 0.3.7

* Handle buffering more consistently across different Python versions.

## 0.3.6

* LocalShell: Add support for Python 3. Since paramiko is currently unsupported
  on Python 3, use the package "spur.local" rather than "spur".

## 0.3.5

* SshShell: Use "command -v" instead of "which" for better POSIX compliance.
* SshShell: Skip blank lines when expecting echoed return code.

## 0.3.4

* LocalShell: Use Popen.wait instead of Popen.poll to get return code of local
  process to ensure process has exited.

## 0.3.3

* Make username argument to SshShell optional. Closes GitHub issue #4.

## 0.3.2

* Include original error and original traceback on spur.ssh.ConnectionError
* Add experimental use_pty argument for run and spawn. Use at your own risk!

## 0.3.1

* spur.NoSuchCommandError is now raised if the command passed to run
  or spawn doesn't exist

## 0.3.0

* Change default behaviour to raise an error when a host key is missing.
* Allow selection of behaviour when a host key is missing by adding
  host_key_missing argument to SshShell constructor.

## 0.2.4

* Catch EOFError and wrap it in spur.ssh.ConnectionError when opening SSH
  session


spur/ssh.py
from __future__ import unicode_literals
from __future__ import absolute_import

import subprocess
import os
import os.path
import shutil
import contextlib
import uuid
import socket
import traceback
import sys
import io

import paramiko

from .tempdir import create_temporary_dir
from .files import FileOperations
from . import results
from .io import IoHandler, Channel
from .errors import NoSuchCommandError, CommandInitializationError, CouldNotChangeDirectoryError


_ONE_MINUTE = 60


class ConnectionError(Exception):
    pass


class UnsupportedArgumentError(Exception):
    pass


class AcceptParamikoPolicy(paramiko.MissingHostKeyPolicy):
    def missing_host_key(self, client, hostname, key):
        return


class MissingHostKey(object):
    raise_error = paramiko.RejectPolicy()
    warn = paramiko.WarningPolicy()
    auto_add = paramiko.AutoAddPolicy()
    accept = AcceptParamikoPolicy()


class MinimalShellType(object):
    supports_which = False

    def generate_run_command(self, command_args, store_pid,
            cwd=None, update_env={}, new_process_group=False):

        if store_pid:
            raise self._unsupported_argument_error("store_pid")

        if cwd is not None:
            raise self._unsupported_argument_error("cwd")

        if update_env:
            raise self._unsupported_argument_error("update_env")

        if new_process_group:
            raise self._unsupported_argument_error("new_process_group")

        return " ".join(map(escape_sh, command_args))


    def _unsupported_argument_error(self, name):
        return UnsupportedArgumentError("'{0}' is not supported when using a minimal shell".format(name))


class ShShellType(object):
    supports_which = True

    def generate_run_command(self, command_args, store_pid,
            cwd=None, update_env={}, new_process_group=False):
        commands = []

        if store_pid:
            commands.append("echo $$")

        if cwd is not None:
            commands.append("cd {0} 2>&1 || {{ echo '\n'spur-cd: $?; exit 1; }}".format(escape_sh(cwd)))
            commands.append("echo '\n'spur-cd: 0")

        update_env_commands = [
            "export {0}={1}".format(key, escape_sh(value))
            for key, value in _iteritems(update_env)
        ]
        commands += update_env_commands
        which_commands = " || ".join(self._generate_which_commands(command_args[0]))
        which_commands = "{ { " + which_commands + "; } && echo 0; } || { echo $?; exit 1; }"
        commands.append(which_commands)

        command = " ".join(map(escape_sh, command_args))
        command = "exec {0}".format(command)
        if new_process_group:
            command = "setsid {0}".format(command)
        commands.append(command)
        return "; ".join(commands)

    def _generate_which_commands(self, command):
        which_commands = ["command -v {0}", "which {0}"]
        return (
            self._generate_which_command(which, command)
            for which in which_commands
        )

    def _generate_which_command(self, which, command):
        return which.format(escape_sh(command)) + " > /dev/null 2>&1"


class ShellTypes(object):
    minimal = MinimalShellType()
    sh = ShShellType()


class SshShell(object):
    def __init__(self,
            hostname,
            username=None,
            password=None,
            port=None,
            private_key_file=None,
            connect_timeout=None,
            missing_host_key=None,
            shell_type=None,
            look_for_private_keys=True,
            load_system_host_keys=True,
            sock=None,
            stdout=None,
            stderr=None):

        if connect_timeout is None:
            connect_timeout = _ONE_MINUTE

        if port is None:
            port = 22

        if shell_type is None:
            shell_type = ShellTypes.sh

        self._hostname = hostname
        self._port = port
        self._username = username
        self._password = password
        self._private_key_file = private_key_file
        self._client = None
        self._connect_timeout = connect_timeout
        self._look_for_private_keys = look_for_private_keys
        self._load_system_host_keys = load_system_host_keys
        self._closed = False
        self._sock = sock
        self._stdout = stdout
        self._stderr = stderr

        if missing_host_key is None:
            self._missing_host_key = MissingHostKey.raise_error
        else:
            self._missing_host_key = missing_host_key

        self._shell_type = shell_type

    def __enter__(self):
        return self

    def __exit__(self, *args):
        self.close()

    def close(self):
        self._closed = True
        if self._client is not None:
            self._client.close()

    def run(self, *args, **kwargs):
        return self.spawn(*args, **kwargs).wait_for_result()

    def spawn(self, command, *args, **kwargs):
        stdout = kwargs.pop("stdout", None)
        if stdout is None:
            stdout = self._stdout
        stderr = kwargs.pop("stderr", None)
        if stderr is None:
            stderr = self._stderr
        allow_error = kwargs.pop("allow_error", False)
        store_pid = kwargs.pop("store_pid", False)
        use_pty = kwargs.pop("use_pty", False)
        encoding = kwargs.pop("encoding", None)
        cwd = kwargs.get('cwd')
        command_in_cwd = self._shell_type.generate_run_command(command, *args, store_pid=store_pid, **kwargs)
        try:
            channel = self._get_ssh_transport().open_session()
        except EOFError as error:
            raise self._connection_error(error)
        if use_pty:
            channel.get_pty()
        channel.exec_command(command_in_cwd)

        process_stdout = channel.makefile('rb')

        if store_pid:
            pid = _read_int_initialization_line(process_stdout)

        if cwd is not None:
            cd_output = []
            while True:
                line = process_stdout.readline()
                if line.startswith(b"spur-cd: "):
                    if line.strip() == b"spur-cd: 0":
                        break
                    else:
                        raise CouldNotChangeDirectoryError(cwd, b"".join(cd_output))
                else:
                    cd_output.append(line)

        if self._shell_type.supports_which:
            which_return_code = _read_int_initialization_line(process_stdout)

            if which_return_code != 0:
                raise NoSuchCommandError(command[0])

        process = SshProcess(
            channel,
            allow_error=allow_error,
            process_stdout=process_stdout,
            stdout=stdout,
            stderr=stderr,
            encoding=encoding,
            shell=self,
        )
        if store_pid:
            process.pid = pid

        return process

    @contextlib.contextmanager
    def temporary_dir(self):
        result = self.run(["mktemp", "--directory"], encoding="ascii")
        temp_dir = result.output.strip()
        try:
            yield temp_dir
        finally:
            self.run(["rm", "-rf", temp_dir])

    def upload_dir(self, local_dir, remote_dir, ignore):
        with create_temporary_dir() as temp_dir:
            content_tarball_path = os.path.join(temp_dir, "content.tar.gz")
            content_path = os.path.join(temp_dir, "content")
            shutil.copytree(local_dir, content_path, ignore=shutil.ignore_patterns(*ignore))
            subprocess.check_call(
                ["tar", "czf", content_tarball_path, "content"],
                cwd=temp_dir
            )
            with self._connect_sftp() as sftp:
                remote_tarball_path = "/tmp/{0}.tar.gz".format(uuid.uuid4())
                sftp.put(content_tarball_path, remote_tarball_path)
                self.run(["mkdir", "-p", remote_dir])
                self.run([
                    "tar", "xzf", remote_tarball_path,
                    "--strip-components", "1", "--directory", remote_dir
                ])

                sftp.remove(remote_tarball_path)

    def open(self, name, mode="r"):
        sftp = self._open_sftp_client()
        sftp_file = SftpFile(sftp, sftp.open(name, mode), mode)

        if "b" not in mode:
            sftp_file = io.TextIOWrapper(sftp_file)

        return sftp_file

    @property
    def files(self):
        return FileOperations(self)

    def _get_ssh_transport(self):
        try:
            return self._connect_ssh().get_transport()
        except (socket.error, paramiko.SSHException, EOFError) as error:
            raise self._connection_error(error)

    def _connect_ssh(self):
        if self._client is None:
            if self._closed:
                raise RuntimeError("Shell is closed")
            client = paramiko.SSHClient()
            if self._load_system_host_keys:
                client.load_system_host_keys()
            client.set_missing_host_key_policy(self._missing_host_key)
            client.connect(
                hostname=self._hostname,
                port=self._port,
                username=self._username,
                password=self._password,
                key_filename=self._private_key_file,
                look_for_keys=self._look_for_private_keys,
                timeout=self._connect_timeout,
                sock=self._sock
            )
            self._client = client
        return self._client

    @contextlib.contextmanager
    def _connect_sftp(self):
        sftp = self._open_sftp_client()
        try:
            yield sftp
        finally:
            sftp.close()

    def _open_sftp_client(self):
        return self._get_ssh_transport().open_sftp_client()

    def _connection_error(self, error):
        connection_error = ConnectionError(
            "Error creating SSH connection\n" +
            "Original error: {0}".format(error)
        )
        connection_error.original_error = error
        connection_error.original_traceback = traceback.format_exc()
        return connection_error


def _read_int_initialization_line(output_file):
    while True:
        line = output_file.readline().strip()
        if line:
            try:
                return int(line)
            except ValueError:
                raise CommandInitializationError(line)


class SftpFile(object):
    def __init__(self, sftp, file, mode):
        self._sftp = sftp
        self._file = file
        self._mode = mode

    def __getattr__(self, key):
        return getattr(self._file, key)

    def close(self):
        try:
            self._file.close()
        finally:
            self._sftp.close()

    def readable(self):
        return "r" in self._mode or "+" in self._mode

    def writable(self):
        return "w" in self._mode or "+" in self._mode or "a" in self._mode

    def seekable(self):
        return True


    def __enter__(self):
        return self

    def __exit__(self, *args):
        self.close()


def escape_sh(value):
    return "'" + value.replace("'", "'\\''") + "'"


class SshProcess(object):
    def __init__(self, channel, allow_error, process_stdout, stdout, stderr, encoding, shell):
        self._channel = channel
        self._allow_error = allow_error
        self._stdin = channel.makefile('wb')
        self._stdout = process_stdout
        self._stderr = channel.makefile_stderr('rb')
        self._shell = shell
        self._result = None

        self._io = IoHandler([
            Channel(self._stdout, stdout),
            Channel(self._stderr, stderr),
        ], encoding=encoding)

    def is_running(self):
        return not self._channel.exit_status_ready()

    def stdin_write(self, value):
        self._channel.sendall(value)

    def send_signal(self, signal):
        self._shell.run(["kill", "-{0}".format(signal), str(self.pid)])

    def wait_for_result(self):
        if self._result is None:
            self._result = self._generate_result()

        return self._result

    def _generate_result(self):
        output, stderr_output = self._io.wait()
        return_code = self._channel.recv_exit_status()

        return results.result(
            return_code,
            self._allow_error,
            output,
            stderr_output
        )


if sys.version_info[0] < 3:
    _iteritems = lambda d: d.iteritems()
else:
    _iteritems = lambda d: d.items()
